File: ds/dlist.go
package ds

import (

	"fmt"
)

type DNode struct {
	Value string
	Next  *DNode
	Prev  *DNode
}

type DList struct {
	Head *DNode
	Tail *DNode
}

func NewDList() *DList {
	return &DList{}
}

func (d *DList) AddHead(val string) {
	node := &DNode{Value: val, Next: d.Head, Prev: nil}
	if d.Head != nil {
		d.Head.Prev = node
	} else {
		d.Tail = node
	}
	d.Head = node
}

func (d *DList) AddTail(val string) {
	node := &DNode{Value: val, Next: nil, Prev: d.Tail}
	if d.Tail != nil {
		d.Tail.Next = node
	} else {
		d.Head = node
	}
	d.Tail = node
}

func (d *DList) AddBefore(target, val string) {
	if d.Head == nil {
		return
	}
	for cur := d.Head; cur != nil; cur = cur.Next {
		if cur.Value == target {
			node := &DNode{Value: val, Next: cur, Prev: cur.Prev}
			if cur.Prev != nil {
				cur.Prev.Next = node
			} else {
				d.Head = node
			}
			cur.Prev = node
			return
		}
	}
}

func (d *DList) AddAfter(target, val string) {
	if d.Head == nil {
		return
	}
	for cur := d.Head; cur != nil; cur = cur.Next {
		if cur.Value == target {
			node := &DNode{Value: val, Next: cur.Next, Prev: cur}
			if cur.Next != nil {
				cur.Next.Prev = node
			} else {
				d.Tail = node
			}
			cur.Next = node
			return
		}
	}
}

func (d *DList) DelHead() {
	if d.Head == nil {
		return
	}
	d.Head = d.Head.Next
	if d.Head != nil {
		d.Head.Prev = nil
	} else {
		d.Tail = nil
	}
}

func (d *DList) DelTail() {
	if d.Tail == nil {
		return
	}
	d.Tail = d.Tail.Prev
	if d.Tail != nil {
		d.Tail.Next = nil
	} else {
		d.Head = nil
	}
}

func (d *DList) findNode(val string) *DNode {
	for cur := d.Head; cur != nil; cur = cur.Next {
		if cur.Value == val {
			return cur
		}
	}
	return nil
}

func (d *DList) DelByVal(val string) {
	cur := d.findNode(val)
	if cur == nil {
		return
	}
	if cur.Prev != nil {
		cur.Prev.Next = cur.Next
	} else {
		d.Head = cur.Next
	}
	if cur.Next != nil {
		cur.Next.Prev = cur.Prev
	} else {
		d.Tail = cur.Prev
	}
}

func (d *DList) Contains(val string) bool {
	return d.findNode(val) != nil
}

func (d *DList) ReadForward() {
	if d.Head == nil {
		fmt.Println("Список пуст")
		return
	}
	fmt.Print("Список вперед: ")
	for cur := d.Head; cur != nil; cur = cur.Next {
		fmt.Printf(" %s ", cur.Value)
	}
	fmt.Println()
}

func (d *DList) ReadBackward() {
	if d.Tail == nil {
		fmt.Println("Список пуст")
		return
	}
	fmt.Print("Список назад: ")
	for cur := d.Tail; cur != nil; cur = cur.Prev {
		fmt.Printf("\"%s\" ", cur.Value)
	}
	fmt.Println()
}

func (d *DList) DelAfterValue(val string) {
	targetNode := d.findNode(val)
	if targetNode == nil || targetNode.Next == nil {
		return
	}
	d.DelByVal(targetNode.Next.Value)
}

func (d *DList) DelBeforeValue(val string) {
	targetNode := d.findNode(val)
	if targetNode == nil || targetNode.Prev == nil {
		return
	}
	d.DelByVal(targetNode.Prev.Value)
}

func (d *DList) toSlice() []string {
	out := []string{}
	for cur := d.Head; cur != nil; cur = cur.Next {
		out = append(out, cur.Value)
	}
	return out
}


func (d *DList) SaveToFile(filename string) error {
    return SaveStringsText(filename, d.toSlice())
}

func (d *DList) LoadFromFile(filename string) error {
    vals, err := LoadStringsText(filename)
    if err != nil { return err }
    d.Head = nil
    d.Tail = nil
    for _, v := range vals {
        d.AddTail(v)
    }
    return nil
}

func (d *DList) SaveToBinaryFile(filename string) error {
    return SaveStringsBinary(filename, d.toSlice())
}

func (d *DList) LoadFromBinaryFile(filename string) error {
    vals, err := LoadStringsBinary(filename)
    if err != nil { return err }
    d.Head = nil
    d.Tail = nil
    for _, v := range vals {
        d.AddTail(v)
    }
    return nil
}

File: dlist.cpp
#include "dlist.h"
#include <iostream>
#include <fstream>

using namespace std;

DList::DList() : head(nullptr), tail(nullptr) {}

DList::~DList() {
    while (head) delHead();
}

void DList::addHead(const string& val) {
    DNode* newNode = new DNode(val, head, nullptr);
    if (head) head->prev = newNode;
    head = newNode;
    if (!tail) tail = newNode;
}

void DList::addTail(const string& val) {
    DNode* newNode = new DNode(val, nullptr, tail);
    if (tail) tail->next = newNode;
    tail = newNode;
    if (!head) head = newNode;
}

auto DList::findValue(const string& val) const -> DNode* {
    DNode* curr = head;
    while (curr) {
        if (curr->value == val) return curr;
        curr = curr->next;
    }
    return nullptr;
}

void DList::addBefore(const string& target, const string& val) {
    DNode* targetNode = findValue(target);
    if (!targetNode) {
        cout << "Элемент '" << target << "' не найден." << endl;
        return;
    }
    DNode* newNode = new DNode(val, targetNode, targetNode->prev);
    if (targetNode->prev) targetNode->prev->next = newNode;
    else head = newNode;
    targetNode->prev = newNode;
}

void DList::addAfter(const string& target, const string& val) {
    DNode* targetNode = findValue(target);
    if (!targetNode) {
        cout << "Элемент '" << target << "' не найден." << endl;
        return;
    }
    DNode* newNode = new DNode(val, targetNode->next, targetNode);
    if (targetNode->next) targetNode->next->prev = newNode;
    else tail = newNode;
    targetNode->next = newNode;
}

void DList::delHead() {
    if (!head) {
        cout << "Список пуст. Удаление невозможно." << endl;
        return;
    }
    DNode* temp = head;
    head = head->next;
    if (head) head->prev = nullptr;
    else tail = nullptr;
    delete temp;
}

void DList::delTail() {
    if (!tail) {
        cout << "Список пуст. Удаление невозможно." << endl;
        return;
    }
    DNode* temp = tail;
    tail = tail->prev;
    if (tail) tail->next = nullptr;
    else head = nullptr;
    delete temp;
}

void DList::delByVal(const string& val) {
    DNode* targetNode = findValue(val);
    if (!targetNode) {
        cout << "Элемент '" << val << "' не найден." << endl;
        return;
    }
    if (targetNode->prev) targetNode->prev->next = targetNode->next;
    else head = targetNode->next;
    if (targetNode->next) targetNode->next->prev = targetNode->prev;
    else tail = targetNode->prev;
    delete targetNode;
}

auto DList::contains(const string& val) const -> bool {
    return findValue(val) != nullptr;
}

void DList::readForward() const {
    if (!head) {
        cout << "Список пуст." << endl;
        return;
    }
    cout << "Список: ";
    DNode* curr = head;
    while (curr) {
        cout << curr->value;
        if (curr->next) cout << " <-> ";
        curr = curr->next;
    }
    cout << endl;
}

void DList::readBackward() const {
    if (!tail) {
        cout << "Список пуст." << endl;
        return;
    }
    cout << "Список (обратный): ";
    DNode* curr = tail;
    while (curr) {
        cout << curr->value;
        if (curr->prev) cout << " <-> ";
        curr = curr->prev;
    }
    cout << endl;
}

void DList::delAfterValue(const string& val) {
    DNode* targetNode = findValue(val);
    if (!targetNode || !targetNode->next) {
        cout << "Невозможно удалить элемент после '" << val << "'." << endl;
        return;
    }
    DNode* toDelete = targetNode->next;
    targetNode->next = toDelete->next;
    if (toDelete->next) toDelete->next->prev = targetNode;
    else tail = targetNode;
    delete toDelete;
}

void DList::delBeforeValue(const string& val) {
    DNode* targetNode = findValue(val);
    if (!targetNode || !targetNode->prev) {
        cout << "Невозможно удалить элемент перед '" << val << "'." << endl;
        return;
    }
    DNode* toDelete = targetNode->prev;
    targetNode->prev = toDelete->prev;
    if (toDelete->prev) toDelete->prev->next = targetNode;
    else head = targetNode;
    delete toDelete;
}
